external registry ping flake on Jenkins

Error log [here](https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin/6540/console):
```
--- FAIL: TestRegistryClientConnectPulpRegistry-2 (0.31s)
	dockerregistryclient_test.go:47: unable to retrieve image info: error checking for V2 registry at https://registry.access.redhat.com/v2/: Get https://registry.access.redhat.com/v2/: read tcp 209.132.182.63:443: connection reset by peer
FAIL
```

Seen again [here](https://ci.openshift.redhat.com/jenkins/job/merge_pull_requests_origin/4119/console):
```
--- FAIL: TestRegistryClientQuayIOImage-4 (10.00s)
    dockerregistryclient_test.go:132: unexpected error: error checking for V2 registry at https://quay.io/v2/: Get https://quay.io/v2/: net/http: TLS handshake timeout
FAIL
```

--- a/test/integration/dockerregistryclient_test.go
+++ b/test/integration/dockerregistryclient_test.go
@@ -4,10 +4,63 @@ package integration
 
 import (
 	"reflect"
+	"strings"
 	"testing"
+	"time"
 
 	"github.com/openshift/origin/pkg/dockerregistry"
 )
+
+const (
+	pulpRegistryName        = "registry.access.redhat.com"
+	dockerHubV2RegistryName = "index.docker.io"
+	dockerHubV1RegistryName = "registry.hub.docker.com"
+	quayRegistryName        = "quay.io"
+
+	maxRetryCount = 4
+	retryAfter    = time.Millisecond * 500
+)
+
+// retryWhenUnreachable invokes given function several times until it succeeds,
+// returns network unrelated error or maximum number of attempts is reached.
+// Should be used to wrap calls to remote registry to prevent test failures
+// because of short-term outages.
+func retryWhenUnreachable(t *testing.T, f func() error) error {
+	timeout := retryAfter
+	attempt := 0
+	for err := f(); err != nil; err = f() {
+		// Examples of catched error messages:
+		//   Get https://registry.com/v2/: dial tcp registry.com:443: i/o timeout
+		//   Get https://registry.com/v2/: dial tcp: lookup registry.com: no such host
+		//   Get https://registry.com/v2/: dial tcp registry.com:443: getsockopt: connection refused
+		//   Get https://registry.com/v2/: read tcp 127.0.0.1:39849->registry.com:443: read: connection reset by peer
+		//   Get https://registry.com/v2/: net/http: request canceled while waiting for connection
+		//   Get https://registry.com/v2/: net/http: TLS handshake timeout
+		//   the registry "https://registry.com/v2/ " could not be reached
+		reachable := true
+		for _, substr := range []string{
+			"dial tcp",
+			"read tcp",
+			"net/http",
+			"could not be reached",
+		} {
+			if strings.Contains(err.Error(), substr) {
+				reachable = false
+				break
+			}
+		}
+
+		if reachable || attempt >= maxRetryCount {
+			return err
+		}
+
+		t.Logf("registry unreachable \"%v\", retry in %s", err, timeout.String())
+		time.Sleep(timeout)
+		timeout = timeout * 2
+		attempt += 1
+	}
+	return nil
+}
 
 func TestRegistryClientConnect(t *testing.T) {
 	c := dockerregistry.NewClient()
@@ -38,38 +91,59 @@ func TestRegistryClientConnect(t *testing.T) {
 
 func TestRegistryClientConnectPulpRegistry(t *testing.T) {
 	c := dockerregistry.NewClient()
-	conn, err := c.Connect("registry.access.redhat.com", false)
+	conn, err := c.Connect(pulpRegistryName, false)
 	if err != nil {
 		t.Fatal(err)
 	}
-	image, err := conn.ImageByTag("library", "rhel", "latest")
+
+	var image *dockerregistry.Image
+	err = retryWhenUnreachable(t, func() error {
+		image, err = conn.ImageByTag("library", "rhel", "latest")
+		return err
+	})
 	if err != nil {
-		t.Fatalf("unable to retrieve image info: %v", err)
+		t.Fatal(err)
 	}
 	if len(image.ID) == 0 {
 		t.Fatalf("image had no ID: %#v", image)
 	}
 }
 
-func TestRegistryClientV2DockerHub(t *testing.T) {
+func TestRegistryClientDockerHubV2(t *testing.T) {
 	c := dockerregistry.NewClient()
-	conn, err := c.Connect("index.docker.io", false)
+	conn, err := c.Connect(dockerHubV2RegistryName, false)
 	if err != nil {
 		t.Fatal(err)
-	}
-	image, err := conn.ImageByTag("kubernetes", "guestbook", "latest")
-	if err != nil {
-		t.Fatalf("unexpected error: %v", err)
 	}
 
-	// a v1 only path
-	conn, err = c.Connect("registry.hub.docker.com", false)
+	var image *dockerregistry.Image
+	err = retryWhenUnreachable(t, func() error {
+		image, err = conn.ImageByTag("kubernetes", "guestbook", "latest")
+		return err
+	})
 	if err != nil {
 		t.Fatal(err)
 	}
-	image, err = conn.ImageByTag("kubernetes", "guestbook", "latest")
+	if len(image.ID) == 0 {
+		t.Fatalf("image had no ID: %#v", image)
+	}
+}
+
+func TestRegistryClientDockerHubV1(t *testing.T) {
+	c := dockerregistry.NewClient()
+	// a v1 only path
+	conn, err := c.Connect(dockerHubV1RegistryName, false)
 	if err != nil {
-		t.Fatalf("unable to retrieve image info: %v", err)
+		t.Fatal(err)
+	}
+
+	var image *dockerregistry.Image
+	err = retryWhenUnreachable(t, func() error {
+		image, err = conn.ImageByTag("kubernetes", "guestbook", "latest")
+		return err
+	})
+	if err != nil {
+		t.Fatal(err)
 	}
 	if len(image.ID) == 0 {
 		t.Fatalf("image had no ID: %#v", image)
@@ -86,40 +160,54 @@ func TestRegistryClientRegistryNotFound(t *testing.T) {
 	}
 }
 
-func TestRegistryClientImage(t *testing.T) {
-	for _, v2 := range []bool{true, false} {
-		host := "index.docker.io"
-		if !v2 {
-			host = "registry.hub.docker.com"
-		}
-		conn, err := dockerregistry.NewClient().Connect(host, false)
-		if err != nil {
-			t.Fatal(err)
-		}
-
-		if _, err := conn.ImageByTag("openshift", "origin-not-found", "latest"); !dockerregistry.IsRepositoryNotFound(err) && !dockerregistry.IsTagNotFound(err) {
-			t.Errorf("V2=%t: unexpected error: %v", v2, err)
-		}
-
-		image, err := conn.ImageByTag("openshift", "origin", "latest")
-		if err != nil {
-			t.Fatalf("V2=%t: %v", v2, err)
-		}
-		if len(image.ContainerConfig.Entrypoint) == 0 {
-			t.Errorf("V2=%t: unexpected image: %#v", v2, image)
-		}
-		if v2 && !image.PullByID {
-			t.Errorf("V2=%t: should be able to pull by ID %s", v2, image.ID)
-		}
-
-		other, err := conn.ImageByID("openshift", "origin", image.ID)
-		if err != nil {
-			t.Fatal(err)
-		}
-		if !reflect.DeepEqual(other.ContainerConfig.Entrypoint, image.ContainerConfig.Entrypoint) {
-			t.Errorf("V2=%t: unexpected image: %#v", v2, other)
-		}
+func doTestRegistryClientImage(t *testing.T, registry, version string) {
+	conn, err := dockerregistry.NewClient().Connect(registry, false)
+	if err != nil {
+		t.Fatal(err)
 	}
+
+	err = retryWhenUnreachable(t, func() error {
+		_, err := conn.ImageByTag("openshift", "origin-not-found", "latest")
+		return err
+	})
+	if err == nil || (!dockerregistry.IsRepositoryNotFound(err) && !dockerregistry.IsTagNotFound(err)) {
+		t.Errorf("%s: unexpected error: %v", version, err)
+	}
+
+	var image *dockerregistry.Image
+	err = retryWhenUnreachable(t, func() error {
+		image, err = conn.ImageByTag("openshift", "origin", "latest")
+		return err
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(image.ContainerConfig.Entrypoint) == 0 {
+		t.Errorf("%s: unexpected image: %#v", version, image)
+	}
+	if version == "v2" && !image.PullByID {
+		t.Errorf("%s: should be able to pull by ID %s", version, image.ID)
+	}
+
+	var other *dockerregistry.Image
+	err = retryWhenUnreachable(t, func() error {
+		other, err = conn.ImageByID("openshift", "origin", image.ID)
+		return err
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+	if !reflect.DeepEqual(other.ContainerConfig.Entrypoint, image.ContainerConfig.Entrypoint) {
+		t.Errorf("%s: unexpected image: %#v", version, other)
+	}
+}
+
+func TestRegistryClientImageV2(t *testing.T) {
+	doTestRegistryClientImage(t, dockerHubV2RegistryName, "v2")
+}
+
+func TestRegistryClientImageV1(t *testing.T) {
+	doTestRegistryClientImage(t, dockerHubV1RegistryName, "v1")
 }
 
 func TestRegistryClientQuayIOImage(t *testing.T) {
@@ -127,8 +215,11 @@ func TestRegistryClientQuayIOImage(t *testing.T) {
 	if err != nil {
 		t.Fatal(err)
 	}
-
-	if _, err := conn.ImageByTag("coreos", "etcd", "latest"); err != nil {
-		t.Errorf("unexpected error: %v", err)
+	err = retryWhenUnreachable(t, func() error {
+		_, err := conn.ImageByTag("coreos", "etcd", "latest")
+		return err
+	})
+	if err != nil {
+		t.Skip("SKIPPING: unexpected error from quay.io: %v", err)
 	}
 }

